packages/google-cloud-pubsub/src/client/pubsub-schema.client.ts

```
@@ -0,0 +1,166 @@
+import { PubSub, SchemaTypes, SchemaViews } from '@google-cloud/pubsub';
+import { SchemaServiceClient } from '@google-cloud/pubsub/build/src/v1';
+import { readFile } from 'fs/promises';
+import { isEqual } from 'lodash';
+import * as path from 'path';
+
+import {
+  PubsubConfigurationInvalidError,
+  PubsubConfigurationMismatchError,
+} from './pubsub-configuration.error';
+import { PubsubTopicContainer } from './pubsub-topic.container';
+
+export class PubsubSchemaClient {
+  private _schemaClient: SchemaServiceClient | null = null;
+  private _schemaClientPromise: Promise<SchemaServiceClient> | null = null;
+
+  private readonly pubsub: PubSub;
+
+  constructor(pubsub: PubSub) {
+    this.pubsub = pubsub;
+  }
+
+  private get schemaClient(): Promise<SchemaServiceClient> {
+    if (this._schemaClient) {
+      return Promise.resolve(this._schemaClient);
+    }
+
+    if (!this._schemaClientPromise) {
+      this._schemaClientPromise = this._initializeSchemaClient();
+    }
+
+    return this._schemaClientPromise;
+  }
+
+  private async _initializeSchemaClient(): Promise<SchemaServiceClient> {
+    this._schemaClient = await this.pubsub.getSchemaClient();
+
+    return this._schemaClient;
+  }
+
+  public async connectAndValidateSchema(topicContainer: PubsubTopicContainer) {
+    const schemaConfiguration = topicContainer.configuration.schema;
+
+    if (!schemaConfiguration) {
```

Can we log warning that the configuration was not provided?

---

I think it's redundant here because the client is designed to work without a schema. It's normal behavior to work without schema

---

I'm not sure how i feel about introducing a singleton for this, I would rather have the client being a Nestjs Provider where you can either use "async" through `useFactory` or provide the object. This can proxied by our module to generate using the provisioned user configuration. 

Since we're talking about the client, a factory would help when the module is compiling to know whether it is possible to resolve or not whereas here, it would only tell you after the application run and the pubsub execute this block

---

It is strictly an internal helper, not a standalone service. Moving it to a nestjs provider would require exposing the internal PubSub instance to the DI container, which breaks encapsulation.
I think if we keep schemaClient internally, we maintain the decoupling of our client from NestJS, keeping the core logic framework-agnostic.
Also, regarding 'fail fast': if a schema is actually configured, the client attempts to resolve and validate it immediately during .initialize(). Since we await this method inside onApplicationBootstrap, any issues will throw an error right there and stop the application boot process.

---

Fair point!